### ISSUE TITLE
Web auth to authenticate CLI from browser

### ISSUE THREAD

**Issue by Benjamin-Philip** [hexpm/hexpm#1024]
José [mentioned](https://github.com/elixir-lang/elixir/issues/10489#issuecomment-833499522) that you (the hexpm team) are interested in creating a means of authenticating the CLI from the browser. Could I get some more info please ?

**Comment by wojtekmach** [comment]
Hey @Benjamin-Philip, thank you for your interest in this feature.

Currently, we authenticate with Hex using the command line:

```
~% mix hex.user auth
Username: wojtekmach
Account password: ******
```

It has a couple of issues:

- the password prompt is unreliable (https://github.com/erlef/build-and-packaging-wg/issues/1)
- no 2FA support (Hex has 2FA but only on the Web)

By pushing more logic to the server side, we avoid having to duplicate the logic between clients (Hex, rebar3, etc). For authenticating with passwords in particular, using the Web will make it easier to leverage password managers like 1Password, etc.

In a nutshell we want to achieve something like this:

https://user-images.githubusercontent.com/76071/117419324-baa12480-af1c-11eb-971d-6e611f17ee88.mov

(I happened to be already logged-in on GitHub, otherwise I'd have to log in on that screen)

We need to add:
- an endpoint to [hex.pm](https://github.com/hexpm/hexpm)
- a function to [`hex_core`](https://github.com/hexpm/hex_core) to open up the website and after the authentication flow is over, save off the API token we got from the website
- a `mix hex.user auth --web` option to [Hex](https://github.com/hexpm/hex) that uses the function from `hex_core`
- same for [`rebar3_hex`](https://github.com/tsloughter/rebar3_hex)

**Comment by Benjamin-Philip** [comment]
> I hope this gives you some context I guess it is still low on details how exactly it is going to work on the technical side so if you decide to work on it, you'd need to do some experimentation (e.g. with GitHub flow to see how exactly they do it)

Yes it does. Even though it is low on details, I was planning to look at the Github implementation as well. 

>  I think the next step would be to build a proof-of-concept, just adding an endpoint to hexpm and a mix hex.user auth --web task that does the simplest possible thing (I'd skip the 8-letter code for now, etc) and afterwards filling in the missing details and extracting code to hex_core.

Ok will do. I'll notify you by email if I get it ready.

I'm unfamiliar with how hex manages authentication.  So here are some questions:

1. How do you save authenticated users on the website? Do you have a cookie that you update regularly ? (Just out of curiosity)
2. In the web auth flow, what happens if a user is not authenticated even on his browser ? What if he has not even signed up ? Do we redirect him to the Login/Signup page, or do you something else in mind ?
3. Alternatively, do you plan to support logging in using OAuth with a connected google/github/facebook acount ?
4. Is the API token saved by the current authentication method, or do you do something else ?

**Comment by wojtekmach** [comment]
> How do you save authenticated users on the website? Do you have a cookie that you update regularly ? (Just out of curiosity)

We have a cookie, see `AuthHelpers` module.

> In the web auth flow, what happens if a user is not authenticated even on his browser ? What if he has not even signed up ? Do we redirect him to the Login/Signup page, or do you have something else in mind ?

If they are nit logged in they should be redirected to a login screen and upon logging in, they should be presented with the option to finish authorizing the CLI.

If the user doesn’t even have an account we could somehow explicitly handle that but I also think it’s ok if they'd restart the whole flow after signing up.

> Alternatively, do you plan to support logging in using OAuth with a connected google/github/facebook acount ?

I think that is completely orthogonal so let’s not worry about that for now. But no, we have no plans for OAuth integration at the moment.

> Is the API token saved by the current authentication method, or do you do something else ?

Yes, currently upon successful auth, the API token is saved to ~/.hex/hex.config and we’d continue to do so.

**Comment by Benjamin-Philip** [comment]
@wojtekmach , Thanks!

**Comment by Benjamin-Philip** [comment]
@wojtekmach I was browsing through hex and rebar3_hex, and I noticed that a lot of logic is  duplicated ( a good example of which is the API key generation). Why don't you write the tasks in hex_core, and write wrappers for mix and rebar ?
